rfd/0134-cloud-first-deployment.md

```diff
@@ -0,0 +1,213 @@
+because it builds a full suite of artifacts. Cloud users need most of these
+artifacts as well for the agents they run on their own infrastructure, however
+Cloud control plane needs a very limited subset of all artifacts. For this
+reason, the Cloud release and promotion process will be split in 2 parts:
+
+- The workflow that builds artifacts required for the control plane: amd64
+  binaries and Docker images.
+- The workflow that builds remaining artifacts such as deb/rpm packages and
+  Helm charts.
+
+This will allow us to produce rapid releases that affect only control plane
+services (auth, proxy) while still having the ability to push remaining artifacts
+when needed.
+
+Both workflows will run in Github Actions and will reuse existing bits of the
+full self-hosted release pipeline.
+
+### Cloud promotion process
+
+The Cloud release promotion process will be staged, with a different class of
+tenants being upgraded at each stage:
+
+1. Canary
+2. Internal
+3. Trials
+4. Team & Enterprise
+
+The canary tenants will be a set of precreated clusters with some data in them
+which engineers will use for connecting their agents and testing their features.
+Canary clusters will be updated automatically once a Cloud release is cut.
+
+Before promoting the release further, engineers will test any significant change
+in the cloud staging environment and then explicitly sign off that the change is
+ready for deployment by applying a label (e.g. `cloud/verified`) on the master
+PR corresponding to the change. The release manager then will approve the
+further rollout.
+
+Once the change is rolled out to production tenants, engineers will be expected
+to monitor error rates and Cloud metrics relevant to their changes.
+
+Let's now look at some of the edge case scenarios in the release process and
+how we'll be handling them.
+
+### Behavior-incompatible changes
+
+Teleport's component compatibility guarantees become somewhat obsolete in a
+Cloud-first world. Historically, we've been reserving behavior-incompatible
+changes for major releases. The concept of a major release, however, becomes
+non-existent in the Cloud-first world where any change, backwards compatible
+or not, gets deployed to all users within days.
+
+When applied to Cloud release model, and accounted for potential delays in
+agent automatic upgrades, the compatibility requirement becomes:
+
+| Any Cloud release must be always compatible with the oldest version of an
+| agent that's connected to Cloud and enrolled in automatic upgrades.
+
+Compatibility requirements affect control plane (proxy, auth and other services
+we host for our users), agents and clients:
+
+- Control plane. We have full control over it, deploy in tandem and make sure
+  these are always compatible with each other.
+- Agents deployed by users. We always keep backwards compatibilithy between
+  adjacent releases and rely on automatic upgrades to make sure they don't drift.
```

Comment:
What are adjacent releases? Two consecutive cloud releases? If so, do we need to take into account a roll-forward release used to fix a problem just deployed and not count one of those releases? Or is adjacent releases two major releases?

Reply:
I meant consecutive Cloud releases, yes. Consecutive major releases will keep being compatible as well though as we need to preserve compat. guarantees for self-hosted users as well. So tbh in reality I don't think the way we handle backwards compatibility will change much, esp. once we've rolled out automatic upgrades to everyone.